src/main/java/com/vmware/vipclient/i18n/messages/service/PatternService.java

```diff
@@ -30,63 +33,90 @@ public JSONObject getPatternsByCategory(String locale, String category) {
+        if (patterns != null) {
+            logger.debug("Find the pattern for language [{}], region [{}].\n", language, region);// [datetime]
+            // and
+            new PatternCacheService().addPatterns(key, patterns);
+            logger.debug("Pattern is cached for language [{}], region [{}]!\n\n", language, region);
+            return patterns;
+        }
+        if (!LocaleUtility.isDefaultLocale(new Locale(language, region))) {
+            logger.info("Can't find pattern for language [{}] region [{}], look for English pattern as fallback!", language, region);
+            patterns = getPatterns(ConstantsKeys.EN);
         }
         return patterns;
     }
 
     private JSONObject getPatternsFromBundle(String locale) {
         JSONObject patterns = null;
         if (LocaleUtility.isDefaultLocale(locale)) {
-            logger.debug("Got pattern from local bundle!");
+            logger.debug("Look for pattern from local bundle for locale [{}]!", locale);
             patterns = new LocalPatternOpt()
-                    .getPatternsByLocale(ConstantsKeys.EN);
+                    .getEnPatterns(ConstantsKeys.EN);
         } else {
-            if (VIPCfg.getInstance().getMessageOrigin() == DataSourceEnum.VIP) {
-            	patterns = new RemotePatternOpt().getPatternsByLocale(locale);
+            if (VIPCfg.getInstance().getMsgOriginsQueue().get(0) == DataSourceEnum.VIP) {
+                logger.debug("Look for pattern from Singleton Service for locale [{}]!", locale);
+                patterns = new RemotePatternOpt().getPatternsByLocale(locale);
             } else {
-                patterns = new LocalPatternOpt().getPatternsByLocale(locale);
+                logger.debug("Look for pattern from local bundle for locale [{}]!", locale);
+                String normalizedLocale = CommonUtil.getCLDRLocale(locale, localePathMap, localeAliasesMap);
+                logger.debug("Normalized locale for locale [{}] is [{}]", locale, normalizedLocale);
+                patterns = new LocalPatternOpt().getPatternsByLocale(normalizedLocale);
             }
         }
         return patterns;
     }
 
     private JSONObject getPatternsFromBundle(String language, String region) {
         JSONObject patterns = null;
-        if (VIPCfg.getInstance().getMessageOrigin() == DataSourceEnum.VIP) {
-        	patterns = new RemotePatternOpt().getPatternsByLocale(language, region);
+        Locale locale = new Locale(language, region);
+        if (LocaleUtility.isDefaultLocale(locale)) {
+            logger.debug("Look for pattern from local bundle for language [{}], region [{}]!", language, region);
+            patterns = new LocalPatternOpt()
+                    .getEnPatterns(ConstantsKeys.EN);
         } else {
-            patterns = new LocalPatternOpt().getPatternsByLocale(ConstantsKeys.EN);
+            if (VIPCfg.getInstance().getMsgOriginsQueue().get(0) == DataSourceEnum.VIP) {
```

Comment:
(Same comment as above)
Do not use if-else statements to get the message origin from the msgOriginsQueue because the order may change in the future. in other words, .get(0) may not always return DataSourceEnum.VIP even if online mode is supported.

Instead, iterate over msgOriginsQueue. See ComponentService.fetchMessages for reference to do this. Add a createPatternOpt method inside DataSourceEnum.Bundle which will return new LocalePatternOpt(). Add the same method inside DataSourceEnum.VIP which will return new RemotePatternOpt(). Use this DataSourceEnum.createPatternOpt just like in ComponentService.fetchMessages.

If you follow the above, the fallback from online to offline mode (and to any other mode added in the queue in the future) will naturally happen without having to modify this method.